File: app/src/main/java/nolambda/stiker/motionviews/widget/entity/ImageEntity.java
package nolambda.stiker.motionviews.widget.entity;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;

import java.io.Serializable;

import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import nolambda.stiker.motionviews.serialized.BaseEntitySavedState;
import nolambda.stiker.motionviews.serialized.ImageEntitySavedState;
import nolambda.stiker.motionviews.serialized.PaintData;
import nolambda.stiker.motionviews.utils.MotionViewUtils;
import nolambda.stiker.motionviews.viewmodel.Layer;

public class ImageEntity extends MotionEntity {

    @NonNull
    private final Bitmap bitmap;

    @Nullable
    private BitmapProvider bitmapProvider;

    public ImageEntity(@NonNull Layer layer,
                       @NonNull BitmapProvider provider,
                       @IntRange(from = 1) int canvasWidth,
                       @IntRange(from = 1) int canvasHeight) {
        this(layer, provider.getImage(), canvasWidth, canvasHeight);
        this.bitmapProvider = provider;
    }

    public ImageEntity(@NonNull Layer layer,
                       @NonNull Bitmap bitmap,
                       @IntRange(from = 1) int canvasWidth,
                       @IntRange(from = 1) int canvasHeight) {
        super(layer, canvasWidth, canvasHeight);

        this.bitmap = bitmap;
        float width = bitmap.getWidth();
        float height = bitmap.getHeight();

        float widthAspect = 1.0F * canvasWidth / width;
        float heightAspect = 1.0F * canvasHeight / height;
        // fit the smallest size
        holyScale = Math.min(widthAspect, heightAspect);

        // initial position of the entity
        srcPoints[0] = 0;
        srcPoints[1] = 0;
        srcPoints[2] = width;
        srcPoints[3] = 0;
        srcPoints[4] = width;
        srcPoints[5] = height;
        srcPoints[6] = 0;
        srcPoints[7] = height;
        srcPoints[8] = 0;
        srcPoints[8] = 0;
    }

    @Override
    public void drawContent(@NonNull Canvas canvas, @Nullable Paint drawingPaint) {
        canvas.drawBitmap(bitmap, matrix, drawingPaint);
    }

    @Override
    public int getWidth() {
        return bitmap.getWidth();
    }

    @Override
    public int getHeight() {
        return bitmap.getHeight();
    }

    @Override
    public void release() {
        if (!bitmap.isRecycled()) {
            bitmap.recycle();
        }
    }

    @Override
    public BaseEntitySavedState serialize() {
        if (bitmapProvider == null) {
            throw new IllegalStateException("BitmapProvider must be used to serialize ImageEntity");
        }
        return new ImageEntitySavedState(
                getLayer(),
                MotionViewUtils.getMatrixValues(matrix),
                holyScale,
                canvasWidth,
                canvasHeight,
                new PaintData(getBorderPaint().getStrokeWidth(), getBorderPaint().getColor()),
                bitmapProvider
        );
    }

    public interface BitmapProvider extends Serializable {
        Bitmap getImage();
    }

    public static class FileBitmapProvider implements BitmapProvider {

        private final String filePath;

        public FileBitmapProvider(String filePath) {
            this.filePath = filePath;
        }

        @Override
        public Bitmap getImage() {
            return BitmapFactory.decodeFile(filePath);
        }
    }

    public static class ResourceBitmapProvider implements BitmapProvider {
        private final int resourceId;
        private final Context context;

        public ResourceBitmapProvider(Context context, int resourceId) {
            this.resourceId = resourceId;
            this.context = context;
        }

        @Override
        public Bitmap getImage() {
            return BitmapFactory.decodeResource(context.getResources(), resourceId);
        }
    }
}

File: app/src/main/java/nolambda/stiker/motionviews/widget/entity/ImageEntity.kt
package nolambda.stiker.motionviews.widget.entity

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import androidx.annotation.DrawableRes
import androidx.annotation.IntRange
import nolambda.stiker.motionviews.serialized.BaseEntitySavedState
import nolambda.stiker.motionviews.serialized.ImageEntitySavedState
import nolambda.stiker.motionviews.serialized.PaintData
import nolambda.stiker.motionviews.utils.MotionViewUtils
import nolambda.stiker.motionviews.viewmodel.Layer

import java.io.Serializable

class ImageEntity(
    layer: Layer,
    private val bitmapProvider: BitmapProvider,
    @IntRange(from = 1) canvasWidth: Int,
    @IntRange(from = 1) canvasHeight: Int
) : MotionEntity(layer, canvasWidth, canvasHeight) {

    private val bitmap by lazy { bitmapProvider.image }

    init {
        val width = bitmap.width.toFloat()
        val height = bitmap.height.toFloat()

        val widthAspect = 1.0f * canvasWidth / width
        val heightAspect = 1.0f * canvasHeight / height
        // fit the smallest size
        holyScale = Math.min(widthAspect, heightAspect)

        // initial position of the entity
        srcPoints[0] = 0f
        srcPoints[1] = 0f
        srcPoints[2] = width
        srcPoints[3] = 0f
        srcPoints[4] = width
        srcPoints[5] = height
        srcPoints[6] = 0f
        srcPoints[7] = height
        srcPoints[8] = 0f
        srcPoints[8] = 0f
    }

    public override fun drawContent(canvas: Canvas, drawingPaint: Paint?) {
        canvas.drawBitmap(bitmap, matrix, drawingPaint)
    }

    override val width: Int get() = bitmap.width

    override val height: Int get() = bitmap.height

    override fun release() {
        if (!bitmap.isRecycled) {
            bitmap.recycle()
        }
    }

    override fun serialize(): BaseEntitySavedState? {
        return ImageEntitySavedState(
            layer,
            MotionViewUtils.getMatrixValues(matrix),
            holyScale,
            canvasWidth,
            canvasHeight,
            PaintData(borderPaint.strokeWidth, borderPaint.color),
            bitmapProvider
        )
    }

    interface BitmapProvider : Serializable {
        val image: Bitmap
    }

    class FileBitmapProvider(private val filePath: String) : BitmapProvider {
        override val image: Bitmap
            get() = BitmapFactory.decodeFile(filePath)
    }

    class ResourceBitmapProvider(
        private val context: Context,
        @DrawableRes private val resourceId: Int
    ) : BitmapProvider {
        override val image: Bitmap
            get() = BitmapFactory.decodeResource(context.resources, resourceId)
    }
}
